### ISSUE TITLE
Should we continue verifying addresses of GC things?

### ISSUE THREAD

**Issue by michaelwu** [servo/servo#6383]
This is a follow up to https://github.com/servo/servo/pull/6150

With the old SM, pointers from layout were checked to ensure they point to valid GC things. With exact rooting, those checks aren't necessary in SM anymore since there is no stack scanning. I believe that verifying addresses is also ineffective now that we have moving objects. GC things are stored in arenas which pack similar sized objects together. It seems likely for a GC thing to be destroyed and have a similar object dropped in its place.

Instead of verifying addresses, I think we should make sure addresses from layout are always valid or have a way to translate them to a valid address. There's two options that come to mind:

1. Rooting all objects layout cares about and have layout use handles/references to the rooted location.
2. Keep track of where objects move to and when they're deleted. Hooks in JSClass are available to do this.

CC @pcwalton @jdm

**Comment by pcwalton** [comment]
I'd lean toward #2 since it seems like it'll be easier to make thread safe.

**Comment by nox** [comment]
I don't understand what this issue is about. Doesn't `LayoutDom<T>` store a stable address that doesn't move under GC anyway?

**Comment by jdm** [comment]
This is talking about the JS reflectors, not the associated native objects.

**Comment by nox** [comment]
@jdm Where are these addresses verified?

**Comment by jdm** [comment]
https://github.com/servo/servo/blob/master/components/script/dom/node.rs#L986-L989 is the code under discussion.

**Comment by nox** [comment]
Maybe I'm being stupid, but aren't we just sending around the address of the Node itself, just as with `Trusted<T>`?

**Comment by jdm** [comment]
No, we send the JS reflector, hence why we cast the uint to a *mut JSObject in the code that I just linked.

**Comment by sagudev** [comment]
Hit in https://github.com/servo/servo/issues/47577, why don't we use addresses of native (DOM) objects which are not movable?
